vault/plugin_catalog.go

```diff
@@ -461,6 +461,124 @@ func (c *PluginCatalog) getBackendPluginType(ctx context.Context, pluginRunner *
+	merr := &multierror.Error{}
+	// Attempt to run as backend plugin
+	config := pluginutil.PluginClientConfig{
+		Name:            pluginRunner.Name,
+		PluginSets:      backendplugin.PluginSet,
+		HandshakeConfig: backendplugin.HandshakeConfig,
+		Logger:          log.NewNullLogger(),
+		IsMetadataMode:  false,
+		AutoMTLS:        true,
+	}
+
+	var client logical.Backend
+	// First, attempt to run as backend V5 plugin
+	c.logger.Debug("attempting to load backend plugin", "name", pluginRunner.Name)
+	pc, err := c.newPluginClient(ctx, pluginRunner, config)
+	if err == nil {
+		// we spawned a subprocess, so make sure to clean it up
+		defer c.cleanupExternalPlugin(pluginRunner.Name, pc.id)
+
+		// dispense the plugin so we can get its version
+		client, err = backendplugin.Dispense(pc.ClientProtocol, pc)
+		if err == nil {
+			c.logger.Debug("successfully dispensed v5 backend plugin", "name", pluginRunner.Name)
+
+			err = client.Setup(ctx, &logical.BackendConfig{})
+			if err != nil {
+				return logical.EmptyPluginVersion, nil
+			}
+			if versioner, ok := client.(logical.PluginVersioner); ok {
+				return versioner.PluginVersion(), nil
+			}
+			return logical.EmptyPluginVersion, nil
+		}
+		merr = multierror.Append(merr, fmt.Errorf("failed to dispense plugin as backend v5: %w", err))
+	}
+	c.logger.Debug("failed to dispense v5 backend plugin", "name", pluginRunner.Name, "error", err)
+	config.AutoMTLS = false
+	config.IsMetadataMode = true
+	// attempt to run as a v4 backend plugin
+	client, err = backendplugin.NewPluginClient(ctx, nil, pluginRunner, log.NewNullLogger(), true)
+	if err != nil {
+		merr = multierror.Append(merr, fmt.Errorf("failed to dispense v4 backend plugin: %w", err))
+		c.logger.Debug("failed to dispense v4 backend plugin", "name", pluginRunner.Name, "error", merr)
+		return logical.EmptyPluginVersion, merr.ErrorOrNil()
+	}
+	c.logger.Debug("successfully dispensed v4 backend plugin", "name", pluginRunner.Name)
+	defer client.Cleanup(ctx)
+
+	err = client.Setup(ctx, &logical.BackendConfig{})
+	if err != nil {
+		return logical.EmptyPluginVersion, err
+	}
+	if versioner, ok := client.(logical.PluginVersioner); ok {
+		return versioner.PluginVersion(), nil
+	}
+	return logical.EmptyPluginVersion, nil
+}
+
+// getDatabaseRunningVersion returns the version reported by a database plugin
+func (c *PluginCatalog) getDatabaseRunningVersion(ctx context.Context, pluginRunner *pluginutil.PluginRunner) (logical.PluginVersion, error) {
+	merr := &multierror.Error{}
+	config := pluginutil.PluginClientConfig{
+		Name:            pluginRunner.Name,
+		PluginSets:      v5.PluginSets,
+		PluginType:      consts.PluginTypeDatabase,
+		Version:         pluginRunner.Version,
+		HandshakeConfig: v5.HandshakeConfig,
+		Logger:          log.Default(),
+		IsMetadataMode:  true,
+		AutoMTLS:        true,
+	}
+
+	// Attempt to run as database V5 or V6 multiplexed plugin
```

Comment:
Is there a V6 now? I only see v5 and v4 referenced below.

Reply:
I copied this comment from elsewhere in the code. :)